sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala

```diff
@@ -296,6 +296,46 @@ case class UpdateTable(
   override def children: Seq[LogicalPlan] = table :: Nil
 }
 
+/**
+ * The logical plan of the MERGE INTO command that works for v2 tables.
+ */
+case class MergeIntoTable(
+    targetTable: LogicalPlan,
+    sourceTable: LogicalPlan,
+    mergeCondition: Expression,
+    matchedActions: Seq[MergeAction],
+    notMatchedActions: Seq[MergeAction]) extends Command with SupportsSubquery {
+  override def children: Seq[LogicalPlan] =
+    Seq(targetTable, sourceTable) ++ Nil
+  override def output: Seq[Attribute] = Seq.empty
+  override lazy val resolved: Boolean = expressions.forall(_.resolved) && childrenResolved &&
+      matchedActions.forall(_.resolved) && notMatchedActions.forall(_.resolved)
+}
+
+sealed abstract class MergeAction(
+    condition: Option[Expression]) extends Expression with Unevaluable {
+  override def foldable: Boolean = false
+  override def nullable: Boolean = false
+  override def dataType: DataType = null
+  override def children: Seq[Expression] = condition.toSeq
+}
+
+case class DeleteAction(deleteCondition: Option[Expression]) extends MergeAction(deleteCondition)
```

Comment:
nit: `case class DeleteAction(condition: Option[Expression])` since we already have delete in the class name

Reply:
done